DataManagementSystem/scripts/dirac-dms-directory-sync.py

```diff
@@ -0,0 +1,552 @@
+  """
+  Wrapper for downloading files
+  """
+  if delete:
+    for filename in result['Value']['Delete']['Files']:
+      res = removeLocalFile(dest_dir+"/"+ filename)
+      if not res['OK']:
+        gLogger.fatal('Deleting of file: ' + filename + ' -X- [FAILED] ' + res['Message'])
+        DIRAC.exit( 1 )
+      else:
+        gLogger.notice("Deleting "+ filename + " -> [DONE]")
+
+    for directoryname in result['Value']['Delete']['Directories']:
+      res = removeLocaDirectory( dest_dir + "/" + directoryname )
+      if not res['OK']:
+        gLogger.fatal('Deleting of directory: ' + directoryname + ' -X- [FAILED] ' + res['Message'])
+        DIRAC.exit( 1 )
+      else:
+        gLogger.notice("Deleting "+ directoryname + " -> [DONE]")
+
+  for directoryname in result['Value']['Create']['Directories']:
+    res = createLocalDirectory( dest_dir+"/"+ directoryname )
+    if not res['OK']:
+      gLogger.fatal('Creation of directory: ' + directoryname + ' -X- [FAILED] ' + res['Message'])
+      DIRAC.exit( 1 )
+    else:
+      gLogger.notice("Creating " + directoryname + " -> [DONE]")
+
+  listOfFiles = result['Value']['Create']['Files']
+  #Chech that we do not have to many threads
+  if nthreads > len(listOfFiles):
+    nthreads = len(listOfFiles)
+
+  if nthreads==0:
+    return S_OK('Upload finished successfully')
+
+  listOfListOfFiles = chunkList( listOfFiles, nthreads )
+  res = runInParallel(arguments=[dm, source_dir, dest_dir], listOfLists=listOfListOfFiles, function=downloadListOfFiles )
+
+  if not res['OK']:
+    return S_ERROR("Download of files failed")
+
+  return S_OK('Upload finished successfully')
+
+def chunkList(alist, nchunks):
+  """
+  Split a list into a list of equaliy sized lists
+  """
+  avg = len(alist) / float(nchunks)
+  out = []
+  last = 0.0
+
+  while last < len(alist):
+    out.append(alist[int(last):int(last + avg)])
+    last += avg
+
+  return out
+
+def downloadListOfFiles(dm, source_dir, dest_dir, listOfFiles, tID):
+  """
+  Wrapper for multithreaded downloading of a list of files
+  """
+  log = gLogger.getSubLogger("[Thread %s] " % tID)
+  threadLine = "[Thread %s]" % tID
+  for filename in listOfFiles:
+    res = downloadRemoteFile(dm, source_dir + "/" + filename, dest_dir + ("/" + filename).rsplit("/", 1)[0])
+    if not res['OK']:
+      log.fatal(threadLine + ' Downloading ' + filename + ' -X- [FAILED] ' + res['Message'])
+      DIRAC.exit( 1 )
+    else:
+      log.notice(threadLine+ " Downloading " + filename + " -> [DONE]")
+
+def runInParallel(arguments, listOfLists, function):
+  """
+  Helper for execution of uploads and downloads in parallel
+  """
+  from multiprocessing import Process
+  processes = []
+  for tID, alist in enumerate(listOfLists):
+    argums = arguments+[alist]+[tID]
+    pro = Process(target=function, args=argums)
+    pro.start()
+    processes.append(pro)
+  for process in processes:
+    process.join()
+
+  for process in processes:
```

Comment:
In DIRAC we have a standard return structure for any bulk operation: S_OK( { "Successful": successful, "Failed": failed } ), where successful and failed are dictionaries with file names as keys and values is either True( or some useful string ) for successful and Error message for the failed. Please, take habit to use this structure.